File: leetcode/33-search.java
class Solution {
    public int search(int[] nums, int target) {
        int len = nums.length;
        int low = 0, high = len-1;
        while(low < high){
            int mid = (low + high) / 2;
            if(nums[mid] > nums[high]) low = mid+1;
            else high = mid;
        }
        int rot = low;
        low = 0; high = len-1;
        while(low <= high){
            int mid = (low + high) / 2;
            int realmid = (mid + rot) % len;
            if(nums[realmid] == target) return realmid;
            if(nums[realmid] < target) low = mid+1;
            else high = mid-1;
        }
        return -1;
    }
}

File: leetcode/33-search.py
class Solution:
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, h = 0, len(nums) - 1
        while l <= h:
            m = l + ((h - l) >> 1)
            if target == nums[m]:
                return m
            elif (nums[m] > nums[l] and target < nums[m] and target >= nums[l]) or (nums[m] < nums[l] and not (target <= nums[h] and target > nums[m])):
                h = m - 1
            else:
                l = m + 1
        return -1
            
        
